drivers/adc/adm1177/iio_adm1177.c

```diff
@@ -0,0 +1,251 @@
+******************************************************************************/
+#include "iio_adm1177.h"
+#include "adm1177.h"
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_delay.h"
+
+static int adm1177_iio_read_raw(void *device, char *buf, uint32_t len,
+				const struct iio_ch_info *channel,
+				intptr_t priv)
+{
+	struct adm1177_iio_dev *iiodev = (struct adm1177_iio_dev *)device;
+	struct adm1177_dev *desc = (struct adm1177_dev *)iiodev->adm1177_dev;
+	int32_t ret, value;
+	uint16_t conv_voltage;
+	uint16_t conv_current;
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = no_os_sign_extend32((uint32_t)conv_voltage, 11);
+
+		break;
+	case IIO_CURRENT:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = no_os_sign_extend32((uint32_t)conv_current, 11);
+
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &value);
+}
+
+static int adm1177_iio_read_scale(void *device, char *buf, uint32_t len,
+				  const struct iio_ch_info *channel,
+				  intptr_t priv)
+{
+	int32_t vals[2];
+	struct adm1177_iio_dev *iio_adm1177;
+	struct adm1177_dev *adm1177;
+
+	if(!device)
+		return -EINVAL;
+
+	iio_adm1177 = device;
+
+	if(!iio_adm1177->adm1177_dev)
+		return -EINVAL;
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		if(adm1177->vrange == ADM1177_VRANGE_LOW)
+			vals[0] = ADM1177_V_FULLSCALE_0;
+		else {
+			vals[1] = ADM1177_V_FULLSCALE_1;
+		}
+		vals[1] = 12;
+		return iio_format_value(buf, len, IIO_VAL_FRACTIONAL_LOG2, 2,
+					vals);
+	case IIO_CURRENT:
+		vals[0] = 105840;
+		vals[1] = 4096 * ADM1177_R_SENSE;
+		return iio_format_value(buf, len, IIO_VAL_FRACTIONAL, 2,
+					vals);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int adm1177_iio_read_samples(void *device, int *buff, uint32_t samples)
+{
+	uint16_t conv_voltage;
+	uint16_t conv_current;
+	uint32_t raw_voltage;
+	uint32_t raw_current;
+	struct adm1177_iio_dev *iio_adm1177;
+	struct adm1177_dev *adm1177;
+
+	if(!device)
+		return -EINVAL;
+
+	iio_adm1177 = (struct adm1177_iio_dev *)device;
+
+	if(!iio_adm1177->adm1177_dev)
+		return -EINVAL;
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	for(uint32_t i = 0; i < samples*2; i++) {
+		adm1177_read(adm1177, &conv_voltage, &conv_current);
+		raw_voltage = (uint32_t)conv_voltage;
+		raw_current = (uint32_t)conv_current;
+		buff[i] = no_os_sign_extend32(raw_voltage, 11);
+		i++;
+		buff[i] = no_os_sign_extend32(raw_current, 11);
+	}
+
+	return samples;
+}
+
+static struct iio_attribute adm1177_iio_voltage_attrs[] = {
+	{
+		.name = "raw",
+		.show = adm1177_iio_read_raw,
+	},
+	{
+		.name = "scale",
+		.show = adm1177_iio_read_scale,
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_attribute adm1177_iio_current_attrs[] = {
```

Comment:
i don't think you need a separate variable since the voltage_attrs is identical

Reply:
That is right.